File: sdk/go/protomcp/tool_test.go
package protomcp_test

import (
	"testing"

	"github.com/msilverblatt/protomcp/sdk/go/protomcp"
)

func TestToolRegistration(t *testing.T) {
	protomcp.ClearRegistry()
	protomcp.Tool("add",
		protomcp.Description("Add two numbers"),
		protomcp.Args(protomcp.IntArg("a"), protomcp.IntArg("b")),
		protomcp.Handler(func(ctx protomcp.ToolContext, args map[string]interface{}) protomcp.ToolResult {
			return protomcp.Result("3")
		}),
	)

	tools := protomcp.GetRegisteredTools()
	if len(tools) != 1 {
		t.Fatalf("expected 1 tool, got %d", len(tools))
	}
	if tools[0].Name != "add" {
		t.Errorf("name = %q, want %q", tools[0].Name, "add")
	}
	if tools[0].Desc != "Add two numbers" {
		t.Errorf("description = %q, want %q", tools[0].Desc, "Add two numbers")
	}
}

func TestArrayArg(t *testing.T) {
	protomcp.ClearRegistry()
	protomcp.Tool("list_items",
		protomcp.Description("List items"),
		protomcp.Args(protomcp.ArrayArg("tags", "string")),
		protomcp.Handler(func(ctx protomcp.ToolContext, args map[string]interface{}) protomcp.ToolResult {
			return protomcp.Result("ok")
		}),
	)

	tools := protomcp.GetRegisteredTools()
	schema := tools[0].InputSchema
	props := schema["properties"].(map[string]interface{})
	tagsProp := props["tags"].(map[string]interface{})
	if tagsProp["type"] != "array" {
		t.Errorf("expected type=array, got %v", tagsProp["type"])
	}
	items := tagsProp["items"].(map[string]interface{})
	if items["type"] != "string" {
		t.Errorf("expected items type=string, got %v", items["type"])
	}
}

func TestObjectArg(t *testing.T) {
	protomcp.ClearRegistry()
	protomcp.Tool("set_config",
		protomcp.Description("Set config"),
		protomcp.Args(protomcp.ObjectArg("config")),
		protomcp.Handler(func(ctx protomcp.ToolContext, args map[string]interface{}) protomcp.ToolResult {
			return protomcp.Result("ok")
		}),
	)

	tools := protomcp.GetRegisteredTools()
	schema := tools[0].InputSchema
	props := schema["properties"].(map[string]interface{})
	configProp := props["config"].(map[string]interface{})
	if configProp["type"] != "object" {
		t.Errorf("expected type=object, got %v", configProp["type"])
	}
}

func TestUnionArg(t *testing.T) {
	protomcp.ClearRegistry()
	protomcp.Tool("process",
		protomcp.Description("Process data"),
		protomcp.Args(protomcp.UnionArg("data", "string", "object")),
		protomcp.Handler(func(ctx protomcp.ToolContext, args map[string]interface{}) protomcp.ToolResult {
			return protomcp.Result("ok")
		}),
	)

	tools := protomcp.GetRegisteredTools()
	schema := tools[0].InputSchema
	props := schema["properties"].(map[string]interface{})
	dataProp := props["data"].(map[string]interface{})
	anyOf := dataProp["anyOf"].([]interface{})
	if len(anyOf) != 2 {
		t.Fatalf("expected 2 anyOf entries, got %d", len(anyOf))
	}
	first := anyOf[0].(map[string]interface{})
	if first["type"] != "string" {
		t.Errorf("expected first type=string, got %v", first["type"])
	}
	second := anyOf[1].(map[string]interface{})
	if second["type"] != "object" {
		t.Errorf("expected second type=object, got %v", second["type"])
	}
}

func TestLiteralArg(t *testing.T) {
	protomcp.ClearRegistry()
	protomcp.Tool("set_mode",
		protomcp.Description("Set mode"),
		protomcp.Args(protomcp.LiteralArg("mode", "fast", "slow", "balanced")),
		protomcp.Handler(func(ctx protomcp.ToolContext, args map[string]interface{}) protomcp.ToolResult {
			return protomcp.Result("ok")
		}),
	)

	tools := protomcp.GetRegisteredTools()
	schema := tools[0].InputSchema
	props := schema["properties"].(map[string]interface{})
	modeProp := props["mode"].(map[string]interface{})
	if modeProp["type"] != "string" {
		t.Errorf("expected type=string, got %v", modeProp["type"])
	}
	enumVals := modeProp["enum"].([]interface{})
	if len(enumVals) != 3 {
		t.Fatalf("expected 3 enum values, got %d", len(enumVals))
	}
	expected := []string{"fast", "slow", "balanced"}
	for i, v := range enumVals {
		if v != expected[i] {
			t.Errorf("enum[%d] = %v, want %v", i, v, expected[i])
		}
	}
}

func TestToolMetadata(t *testing.T) {
	protomcp.ClearRegistry()
	protomcp.Tool("delete_user",
		protomcp.Description("Delete a user account"),
		protomcp.DestructiveHint(true),
		protomcp.Handler(func(ctx protomcp.ToolContext, args map[string]interface{}) protomcp.ToolResult {
			return protomcp.Result("deleted")
		}),
	)

	tools := protomcp.GetRegisteredTools()
	if !tools[0].Destructive {
		t.Error("expected destructive hint")
	}
}

File: sdk/typescript/tests/tool.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { z } from 'zod';
import { tool, getRegisteredTools, clearRegistry } from '../src/tool.js';

describe('tool()', () => {
  beforeEach(() => clearRegistry());

  it('registers a tool with Zod schema', () => {
    const add = tool({
      name: 'add',
      description: 'Add two numbers',
      args: z.object({
        a: z.number().describe('First number'),
        b: z.number().describe('Second number'),
      }),
      handler: (args) => args.a + args.b,
    });

    const tools = getRegisteredTools();
    expect(tools).toHaveLength(1);
    expect(tools[0].name).toBe('add');
    expect(tools[0].description).toBe('Add two numbers');
  });

  it('requires name for arrow function handlers', () => {
    const t = tool({
      description: 'No name',
      args: z.object({ x: z.number() }),
      handler: (args) => args.x,
    });
    // Arrow functions have empty .name, so falls back to tool_N
    expect(t.name).toMatch(/^tool_\d+$/);
  });

  it('generates JSON Schema from Zod', () => {
    tool({
      name: 'search',
      description: 'Search',
      args: z.object({
        query: z.string().describe('Search query'),
        limit: z.number().default(10).describe('Max results'),
      }),
      handler: (args) => [],
    });

    const tools = getRegisteredTools();
    const schema = JSON.parse(tools[0].inputSchemaJson);
    expect(schema.type).toBe('object');
    expect(schema.properties.query.type).toBe('string');
    expect(schema.properties.limit.default).toBe(10);
    expect(schema.required).toContain('query');
    expect(schema.required).not.toContain('limit');
  });

  it('handler is callable', () => {
    const add = tool({
      description: 'Add',
      args: z.object({ a: z.number(), b: z.number() }),
      handler: (args) => args.a + args.b,
    });

    expect(add.handler({ a: 2, b: 3 })).toBe(5);
  });

  it('generates output schema from Zod', () => {
    const t = tool({
      name: 'search',
      description: 'Search',
      args: z.object({ q: z.string() }),
      output: z.object({ title: z.string(), score: z.number() }),
      handler: (args) => ({ title: 'test', score: 0.9 }),
    });
    const tools = getRegisteredTools();
    const def = tools[tools.length - 1];
    expect(def.outputSchemaJson).toBeTruthy();
    const schema = JSON.parse(def.outputSchemaJson);
    expect(schema.properties.title).toBeTruthy();
  });

  it('generates anyOf for union types', () => {
    tool({
      name: 'process',
      description: 'Process data',
      args: z.object({
        data: z.union([z.string(), z.object({}).passthrough()]),
      }),
      handler: (args) => 'done',
    });

    const tools = getRegisteredTools();
    const schema = JSON.parse(tools[tools.length - 1].inputSchemaJson);
    const dataProp = schema.properties.data;
    expect(dataProp.anyOf || dataProp.oneOf).toBeTruthy();
    const variants = dataProp.anyOf || dataProp.oneOf;
    const types = variants.map((v: any) => v.type);
    expect(types).toContain('string');
    expect(types).toContain('object');
  });

  it('generates array with items schema', () => {
    tool({
      name: 'tag_items',
      description: 'Tag items',
      args: z.object({
        tags: z.array(z.string()),
      }),
      handler: (args) => 'done',
    });

    const tools = getRegisteredTools();
    const schema = JSON.parse(tools[tools.length - 1].inputSchemaJson);
    const tagsProp = schema.properties.tags;
    expect(tagsProp.type).toBe('array');
    expect(tagsProp.items.type).toBe('string');
  });

  it('generates enum for z.enum', () => {
    tool({
      name: 'set_mode',
      description: 'Set mode',
      args: z.object({
        mode: z.enum(['fast', 'slow', 'balanced']),
      }),
      handler: (args) => 'done',
    });

    const tools = getRegisteredTools();
    const schema = JSON.parse(tools[tools.length - 1].inputSchemaJson);
    const modeProp = schema.properties.mode;
    expect(modeProp.enum).toEqual(['fast', 'slow', 'balanced']);
  });

  it('generates nullable schema', () => {
    tool({
      name: 'maybe',
      description: 'Maybe null',
      args: z.object({
        value: z.string().nullable(),
      }),
      handler: (args) => 'done',
    });

    const tools = getRegisteredTools();
    const schema = JSON.parse(tools[tools.length - 1].inputSchemaJson);
    const valueProp = schema.properties.value;
    // openApi3 target uses nullable: true
    expect(valueProp.nullable === true || valueProp.anyOf || valueProp.oneOf).toBeTruthy();
  });

  it('includes tool metadata', () => {
    const t = tool({
      name: 'delete_doc',
      description: 'Delete',
      args: z.object({ id: z.string() }),
      title: 'Delete Document',
      destructiveHint: true,
      handler: (args) => 'deleted',
    });
    const tools = getRegisteredTools();
    const def = tools[tools.length - 1];
    expect(def.title).toBe('Delete Document');
    expect(def.destructiveHint).toBe(true);
  });
});
